### PR TITLE
Add Support for STM32L562E-DK Discovery

### PR DESCRIPTION
**Pull request by yestin**
The patches add support for the STM32L562E-DK Discovery board, including the support for the STM32L562QE SoC.

### CODE REVIEW

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright (c) 2020 Yestin Sun
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_STM32L562XX
+
+config SOC
+	default "stm32l562xx"
+
+config NUM_IRQS
+	default 108
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx:12]:
```suggestion
	default 109
```
This value is used as array size

File: boards/arm/stm32l562e_dk/pinmux.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2020 Linaro Limited
```
Comment (erwango) [inline boards/arm/stm32l562e_dk/pinmux.c:2]: Please update

File: boards/arm/stm32l562e_dk/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+| PINMUX    | on-chip    | pinmux                              |
++-----------+------------+-------------------------------------+
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c                                 |
++-----------+------------+-------------------------------------+
+| PWM       | on-chip    | pwm                                 |
++-----------+------------+-------------------------------------+
+| SPI       | on-chip    | spi                                 |
++-----------+------------+-------------------------------------+
```
Comment (erwango) [inline boards/arm/stm32l562e_dk/doc/index.rst:170]: These 3 don't seem supported yet on this board

File: boards/arm/stm32l562e_dk/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+For mode details please refer to `STM32L562E-DK Discovery board User Manual`_.
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+
+- USART_1 TX/RX : PA9/PA10
+- USART_2 TX/RX : PD5/PD6
+- LPUART_1 TX/RX : PB11/PB10
+- I2C_1 SCL/SDA : PB6/PB7
+- SPI_1 SCK/MISO/MOSI/NSS : PG2/PG3/PG4/PG5
```
Comment (erwango) [inline boards/arm/stm32l562e_dk/doc/index.rst:193]: These don't semm to be supported on this board yet

**Review by erwango — CHANGES_REQUESTED** [review]
A couple of nitpicks, otherwise sounds good

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx (hunk 1 above)
Comment (yestin) [reply soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx:12]: OK, I read the datasheets for both STM32L552xx and STM32L562xx. The maximum maskable interrupt channels are both 109. So wonder why it is 108 in the stm32l552xx defconfig?

File: boards/arm/stm32l562e_dk/pinmux.c (hunk 2 above)
Comment (yestin) [reply boards/arm/stm32l562e_dk/pinmux.c]: Thanks for the review. The copyright comment has been updated.

File: boards/arm/stm32l562e_dk/doc/index.rst (hunk 3 above)
Comment (yestin) [reply boards/arm/stm32l562e_dk/doc/index.rst]: The documentation has been updated to remove the item that are not yet supported.

File: boards/arm/stm32l562e_dk/doc/index.rst (hunk 4 above)
Comment (yestin) [reply boards/arm/stm32l562e_dk/doc/index.rst]: The documentation has been updated to remove the items that are not yet supported.

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx (hunk 1 above)
Comment (erwango) [reply soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx]: According to https://raw.githubusercontent.com/zephyrproject-rtos/hal_stm32/master/stm32cube/stm32l5xx/soc/stm32l552xx.h, maximum IRQ number is 107.
There might be a misalignment between ref man and Cube package then.
Even though this is highly theoretical, as max array size might be reached only if all IRQ are configured in same binary, from Zephyr perspective, we're CMSIS file as reference so value for STM32L552xx should be 107+1, until is is potentially fixed.

**Review by erwango — CHANGES_REQUESTED** [review]
LGTM but please squash the fix commits:
as we're using fast-forward merges (all your commits are put on top of master branch), so we expect only cleans commits.

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx (hunk 1 above)
Comment (yestin) [reply soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l562xx]: Ah, I see. Thanks for the explanation.

**Comment by yestin** [comment]
> LGTM but please squash the fix commits:
> as we're using fast-forward merges (all your commits are put on top of master branch), so we expect only cleans commits.

The fixes have been squashed into their original commits. Please help to review if I did it correctly. Thanks.

**Review by erwango — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**